### ISSUE TITLE
URL Parsing/Preview -- Open Graph Meta Tags

### ISSUE THREAD

**Issue by ciaduck** [wishthis/wishthis#188]
## Issue
URL Preview and Parsing does not currently function well.

## Proposed Solution
Implement parsing of Open Graph `<meta>` tags to facilitate auto fill of wish attributes. There may be an existing implementation already, although using an XML parser and following the protocol also seems valid.

## Example
If I were to add this 50mm Lens from B&H, the title is set as "Just a moment...", suggesting that either CURL didn't follow all redirects, or the parsed title is set by a script, and that isn't rendered by CURL.
[Canon EF 50mm f/1.8 STM Lens](https://www.bhphotovideo.com/c/product/1143786-REG/canon_0570c002_ef_50mm_f_1_8_stm.html)
### Relevant meta tags from page source
```
<meta property="og:image:type" content="image/jpeg" data-rh="true">
<meta property="og:image:width" content="200" data-rh="true">
<meta property="og:image:height" content="100" data-rh="true">
<meta property="og:image" content="https://static.bhphoto.com/images/fb/1143786.jpg" data-rh="true">
<meta property="og:image:secure_url" content="https://static.bhphoto.com/images/fb/1143786.jpg" data-rh="true">
<meta property="og:description" content="Canon EF 50mm f/1.8 STM 50mm Lens featuring EF-Mount Lens/Full-Frame Format, Aperture Range: f/1.8 to f/22, 50mm Lens" data-rh="true">
<meta property="og:title" content="Canon EF 50mm f/1.8 STM Lens" data-rh="true">
```

## Additional Resources
https://blog.logrocket.com/open-graph-sharable-social-media-previews/
https://stackoverflow.com/questions/19312942/how-can-i-retrieve-og-meta-attributes-of-a-resource/19313817#19313817

Finally, thank you for your continued and excellent work!

**Comment by grandeljay** [comment]
Hey @ciaduck,

thank you for this report!

If I understood you correctly, this sounds more like a bug. There was a point where wishthis would automatically fetch stuff like Image, title, and description automatically. Are you saying that's no longer working when you supply an url?

There were exceptions (such as Amazon) but I presume they just have bad markup.

**Comment by ciaduck** [comment]
That is correct. For whatever reason it's not working for me. At best, with amazon I get a title. But no URL I've tried will produce image, description, or price.

Could this be related to client side scripting? If this is being fetched on the client side, instead of the server side, I could see that causing issues. I'm using Brave for my browser. I had the impression this was server side though, due to the use of php libcurl.

Thanks.

**Comment by grandeljay** [comment]
I see!

The wish info is fetched from the server side, I'll look into it.

**Comment by grandeljay** [comment]
I've done some debugging and it seems like these services are blocking the crawler which attempts to scrape the info from the wish url.

I'm not sure if/what I can do about that right now.

**Comment by ciaduck** [comment]
I can start trying to look at the crawler code. I just don't have time this week.

Some immediate ideas that come to mind:
- Set a user agent to be something like a real agent string
- Set user agent to be something like a crawler bot (google bot for instance)

I've done a few python based web crawlers in the past, but usually I just reach for `wget` these days.

**Comment by grandeljay** [comment]
Awesome! I've published my branch ([fix/188-prefretch-wish-data](https://github.com/wishthis/wishthis/tree/fix/188-prefretch-wish-data)) with my progress now. So far I've just set the user agent.

**Comment by ciaduck** [comment]
Just a quick update. I'm still looking into this, but it doesn't seem like the issue is with CURL or the PHP Embed library. At first glance it looks like perhaps the wrong tag/data is being assigned to the wrong fields.

I'm working on just having embed dump out all fields to check which ones are populated as expected.

**Comment by grandeljay** [comment]
I hope you find something!

While i was testing I was getting HTTP 403 for each request. I think these providers are annoyed of wishthis scraping their site. It would also explain why this feature has stopped working suddenly, without a change in code.

**Comment by greed77** [comment]
I know this is an old issue, but it's one that I think stands between me and being able to pick wishthis over some alternatives, so I'm going to see if I can be any help :D

I poked at the code a bit, and from what I can tell, it's that more and more ecommerce sites are using something like cloudflare to prevent any scraping at all, not just from wishthis. I did some testing with a few sites and some would pull the meta info, but some wouldn't. I think there may be some super clever ways around it to try to trick the site into showing you the page, but I think that if you're trying to be honest with the site about what you are and what you're doing (which I would respect :P), then a more pragmatic approach might be something like a bookmarklet.

I started out just trying to pull the data out into something that I could copy/paste into wishthis, but then I saw there was an API endpoint for wishlists, so I got something working with that. Is that something that you think might work?

<img width="580" height="1210" alt="Image" src="https://github.com/user-attachments/assets/344135eb-b36c-4d61-bb1b-a8e05427d9e8" />

**Comment by grandeljay** [comment]
Hey @greed77,

> I started out just trying to pull the data out into something that I could copy/paste into wishthis, but then I saw there was an API endpoint for wishlists, so I got something working with that. Is that something that you think might work?

That looks really good! There is an even older issue (#49) asking for a bookmarklet and this seems to solve exactly that. If you're interested in sharing, feel free to open a PR!
